core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -0,0 +1,496 @@
+              connectionName));
+    }
+    if (!instanceMetadata.getBackendType().equals("SECOND_GEN")) {
+      throw new IllegalArgumentException(
+          String.format(
+              "[%s] Connections to Cloud SQL instance not supported - not a Second Generation "
+                  + "instance.",
+              connectionName));
+    }
+
+    // Update the IP addresses and types need to connect with the instance.
+    Map<String, String> ipAddrs = new HashMap<>();
+    for (IpMapping addr : instanceMetadata.getIpAddresses()) {
+      ipAddrs.put(addr.getType(), addr.getIpAddress());
+    }
+    if (ipAddrs.isEmpty()) {
+      throw new IllegalStateException(
+          String.format(
+              "[%s] Unable to connect to Cloud SQL instance: instance does not have an assigned "
+                  + "IP address.",
+              connectionName));
+    }
+
+    // Update the Server CA certificate used to create the SSL connection with the instance.
+    Certificate instanceCaCertificate;
+    try {
+      byte[] certBytes =
+          instanceMetadata.getServerCaCert().getCert().getBytes(StandardCharsets.UTF_8);
+      instanceCaCertificate =
+          CertificateFactory.getInstance("X.509")
+              .generateCertificate(new ByteArrayInputStream(certBytes));
+    } catch (CertificateException ex) {
+      throw new RuntimeException(
+          String.format(
+              "[%s] Unable to parse the server CA certificate for the Cloud SQL instance.",
+              connectionName),
+          ex);
+    }
+
+    return new Metadata(ipAddrs, instanceCaCertificate);
+  }
+
+  /**
+   * Uses the Cloud SQL Admin API to create an ephemeral SSL certificate that is authenticated to
+   * connect the Cloud SQL instance for 60 minutes.
+   */
+  private Certificate fetchEphemeralCertificate() {
+    // Format the public key into a PEM encoded Certificate.
+    StringBuilder publicKeyPemBuilder = new StringBuilder();
+    publicKeyPemBuilder.append("-----BEGIN RSA PUBLIC KEY-----\n");
+    publicKeyPemBuilder.append(
+        Base64.getEncoder()
+            .encodeToString(keyPair.getPublic().getEncoded())
+            .replaceAll("(.{64})", "$1\n"));
+    publicKeyPemBuilder.append("\n");
+    publicKeyPemBuilder.append("-----END RSA PUBLIC KEY-----\n");
+
+    // Use the SQL Admin API to create a new ephemeral certificate.
+    SslCertsCreateEphemeralRequest request =
+        new SslCertsCreateEphemeralRequest().setPublicKey(publicKeyPemBuilder.toString());
+    SslCert response;
+    try {
+      response = apiClient.sslCerts().createEphemeral(projectId, instanceId, request).execute();
+    } catch (IOException ex) {
+      throw addExceptionContext(
+          ex,
+          String.format(
+              "[%s] Failed to create ephemeral certificate for the Cloud SQL instance.",
+              connectionName));
+    }
+
+    // Parse the certificate from the response.
+    Certificate ephemeralCertificate;
```

Comment:
This looks almost the same as the code in `fetchMetadata()`. Can you extract a method that takes a String representing the bytes of a certificate (`response.getCert()`'s return value)?